pointmatcher/DataPointsFilters/utils/octree.h

```diff
@@ -0,0 +1,167 @@
+You can contact the authors at <f dot pomerleau at gmail dot com> and
+<stephane at magnenat dot net>
+
+All rights reserved.
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions are met:
+    * Redistributions of source code must retain the above copyright
+      notice, this list of conditions and the following disclaimer.
+    * Redistributions in binary form must reproduce the above copyright
+      notice, this list of conditions and the following disclaimer in the
+      documentation and/or other materials provided with the distribution.
+    * Neither the name of the <organization> nor the
+      names of its contributors may be used to endorse or promote products
+      derived from this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+DISCLAIMED. IN NO EVENT SHALL ETH-ASL BE LIABLE FOR ANY
+DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+*/
+#pragma once
+
+#include <cstdlib> 
+#include <vector>
+#include "PointMatcher.h"
+
+/*!
+ * \class octree.h
+ * \brief Octree class for DataPoints spatial representation
+ *
+ * \author Mathieu Labussiere (<mathieu dot labu at gmail dot com>)
+ * \date 24/05/2018
+ * \version 0.1
+ *
+ * Octree implementation for decomposing point cloud. 
+ * The current implementation use the data structure PointMatcher<T>::DataPoints. 
+ * It ensures that each node has either 8 or 0 childs. 
+ *
+ * Can create an octree with the 2 following crieterions:
+ *	- max number of data by node
+ *	- max size of a node (or stop when only one or zero data element is available)
+ *
+ * After construction, we can apply a process by creating a Visitor, implementing the following method:
+ *	```cpp
+ *	template<typename T>
+ *	struct Visitor {
+ *		bool operator()(Octree<T>& oc);
+ *	};
+ *	```
+ *
+ * Remark:
+ *	- Current implementation only store the indexes of the points from the pointcloud.
+ *	- Data element are exclusively contained in leaves node.
+ *	- Some leaves node contains no data (to ensure 8 or 0 childs).
+ *
+ */
+template < typename T >
+class Octree 
+{
+public:
+	using PM = PointMatcher<T>;
+	using DP = typename PM::DataPoints; /**/
+	using Id = typename DP::Index; /**/
+	
+	using Data = typename DP::Index; /**/
+	using DataContainer = std::vector<Data>;
+
+private:
+	struct XYZ 
```

Comment:
Isn't that duplicating some functions already provided by Eigen by `Eigen::Matrix<T,3,1>`? It would be nice not to duplicate such basic features.

Reply:
Yes indeed it's kinda a duplicate, but for efficient memory representation I needed a minimal struct for representation a triplet of coordinate. 

However, I tested a version a the octree with `Eigen::Matrix<T,3,1>` instead of the `XYZ` struct. A quick benchmark tends to suggest that the build of the octree with the latter is quicker than with `Eigen` representation, which is `1.89` times slower. 

Any idea how to optimize the use of the `Eigen` matrix (where the XYZ is used) to speedup the build ?

Reply:
You are right, I was missing the optimization flag. With `-O3` the build time is the same for both structure, I'll switch to `Eigen::Matrix<T,3,1>` then. Thanks for pointing that out.